Add preferences UI

vinokurig opened the pull request:
![ezgif-1-36358b34d1](https://user-images.githubusercontent.com/7668752/40655572-6913a62c-634a-11e8-803b-e218ab156816.gif)

svenefftinge commented on packages/core/src/browser/preferences/preference-contribution.ts:
```diff
@@ -16,6 +16,7 @@ export interface PreferenceContribution {
 export const PreferenceSchema = Symbol("PreferenceSchema");
 
 export interface PreferenceSchema {
+    name: string,
```
Instead of adding a group label, we should make use of the convention to use dot-separated property names. 
The tree should be based on the splitted segments. This will allow extensions to contribute preferences to arbitrary groups, too.

svenefftinge commented on packages/preferences/src/browser/preferences-widget-factory.ts:
```diff
@@ -0,0 +1,34 @@
+        const widget = this.widget;
+        widget.scope = options.scope;
+        widget.id = "theia-preferences-container";
+        widget.title.label = 'Prefernces';
```
Preferences

svenefftinge replied:
I think the label should indicate the scope, too. So when coming back later I see whether it is user or workspace scope.

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,252 @@
+    property: PreferenceProperty;
+}
+
+export class PreferencesWidget extends VirtualWidget {
```
Would be good if this could use the tree widget. It would automatically look more aligned with the rest and we can make use of the fast filter, keyboard navigation and decoration support.

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,252 @@
+
+    constructor(@inject(EditorPreferences) protected readonly editorPreferences: EditorPreferences,
+                @inject(EditorManager) protected readonly editorManager: EditorManager,
+                @multiInject(PreferenceSchema) protected readonly preferenceSchema: PreferenceSchema[],
```
Should use `PreferenceSchemaProvider` to obtain the JSON schema.

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,252 @@
+        if (property.type === 'boolean' || property.enum) {
+            valueContainer = this.createEnumItems(preferenceName, property);
+        } else {
+            valueContainer = this.createInputValuePanel(preferenceName, property, onBlur);
```
Please use the standard menu (`ContextMenuRenderer`) which will make sure it looks and feels the same as else where. Also it will make sure that it gets rendered properly (e.g. above the cursor when you are close to the bottom).
Like in VS Code it should propose the options when there is a finite set of alternatives (boolean, enums) otherwise copy the element over and select the value on the right.

vinokurig replied:
Is it possible to set different context menus to different tree items in the same tree?

svenefftinge replied:
Yes, you can directly use the context menu renderer. I.e. without going through the menu registry.

vinokurig replied:
Setting non enum\boolean values directly to file will not give an ability to validate values. Some properties might not have default values, so in that case we wont't know about value type.

svenefftinge replied:
That should be solved by enabling the JSON schema support.

svenefftinge commented on packages/preferences/src/browser/style/prefernces.css:
```diff
@@ -0,0 +1,132 @@
+/*
```
typo in file name

vinokurig replied:
fixed

vinokurig commented:
I will work on that

vinokurig commented:
@svenefftinge Reworked to tree widget:
![screenshot-localhost-3000-2018 06 06-15-54-25](https://user-images.githubusercontent.com/7668752/41039527-5a87de12-69a2-11e8-8e2f-ccb2a00b3cfd.png)

gorkem commented:
@vinokurig Is there a possibility show the descriptions associated with the individual preferences?

vinokurig commented:
Not yet, I'll add a tree decorator for it

vinokurig commented:
@svenefftinge any updates on this?

vinokurig commented:
@svenefftinge I solved the problem with `PreferencesDecoratorService` by injecting `PreferencesDecorator` directly to the service, so the PR is ready to review.

svenefftinge commented on packages/preferences/src/browser/preference-frontend-contribution.ts:
```diff
@@ -24,34 +35,52 @@ export namespace PreferenceCommands {
     @inject(OpenerService) protected readonly openerService: OpenerService;
     @inject(FileSystem) protected readonly filesystem: FileSystem;
 
+    constructor (protected readonly applicationShell: ApplicationShell, protected readonly options: ViewContributionOptions) {
```
Constructor is not necessary. The ApplicationShell should be injected as a property: 

`@inject(ApplicationShell) protected shell: ApplicationShell;`

vinokurig replied:
done

svenefftinge commented on packages/preferences/src/browser/preference-menu-factory.ts:
```diff
@@ -0,0 +1,69 @@
+                menuModel.addNode(new ActionMenuNode({commandId: 'stringProperty'}, this.commands));
+            }
+        }
+        const phosphorCommands = this.createPhosporCommands(menuModel);
```
There is no reason to go through the global command registry. Instead you should directly create a phosphor `CommandRegistry`. That also renders the new APIs in CommandRegistry unnecessary.

vinokurig replied:
`ActionMenuNode` receives commands type of `CommandRegistry` from '@theia/command'

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,142 @@
+        super.onAfterAttach(msg);
+    }
+    protected onCloseRequest(msg: Message): void {
+        this.applicationShell
```
We should compose the two widgets (tree & editor) in one instead of opening the editor on the right panel.

vinokurig replied:
done

svenefftinge commented:
Sorry, I didn't noticed the code was opening two widgets in the first commit already. I would have mentioned it earlier otherwise.

vinokurig commented:
@svenefftinge 
While trying to connect my tree widget and editor widget in one widget I faced with problems:

- I managed to insert the tree widget to my common `VirtualWidget` by calling 'render()` method in the tree widget. It was shown, but without css styles and nodes didn't work:
 ![screenshot-localhost-3000-2018 06 20-17-26-57](https://user-images.githubusercontent.com/7668752/41664824-41b07d1e-74af-11e8-889a-c68ae31aa067.png)
 
- There is no way to add an `EditorWidget` because 'render()` method is not implemented there.

So I wonder if it is possible to gather different widgets in one widget?

svenefftinge commented:
> So I wonder if it is possible to gather different widgets in one widget?

If it is not, we should definitely make it possible. I think it makes a lof of sense to embed an editor in widgets in other cases, too. Would you be able to track down the cause and propose a change for the existing code to support it?

vinokurig commented:
@svenefftinge I've added container widget and inserted the tree and the editor to it. now it looks like this:
![screenshot-localhost-3000-2018 06 22-17-02-49](https://user-images.githubusercontent.com/7668752/41780852-9a292190-763e-11e8-97eb-cb855dbe5703.png)

akosyakov commented on packages/preferences/src/browser/preference-menu-factory.ts:
```diff
@@ -0,0 +1,69 @@
+@injectable()
+export class PreferencesMenuFactory extends BrowserMainMenuFactory {
+
+    private commandsStorage: Command[] = [];
```
you should use `Disposable` pattern instead:
```ts
protected readonly toUnregister = new DisposableCollection();

this.toUregister.push(this.commands.registerCommand(...));

menu.aboutToClose.connect(() => this.toUnregister.dispose());
```

vinokurig replied:
done

akosyakov commented on packages/core/src/common/command.ts:
```diff
@@ -118,6 +118,15 @@ export class CommandRegistry implements CommandService {
+    /**
+     * Unregister command by it's identifier.
+     */
+    unregisterCommand(commandId: string) {
```
please rollback, we use `Disposable` pattern, `registerCommand` return a disposable object which should be disposed to unregister

vinokurig replied:
done

marcdumais-work commented:
Nice feature. I did not go through all the "outdated" comments, so this may already be known: I noticed that if I open the workspace preferences file on its own by double-clicking it in the file navigator, subsequently the "preferences UI"  is missing the editor on the right side. This is fixed after a front-end reload.

vinokurig commented:
@marcdumais-work Thank you for reporting a bug, fixed.

vinokurig commented:
@svenefftinge @marcdumais-work @akosyakov any updates no this?

svenefftinge commented:
- Using the dockpanel doesn't seem right, as we have an odd invisible tabbar on the top.
 - the dirty-ness doesn't work. The preferences editor is just an editor so it should behave like one.
 - the tree should gain focus
 - keyboard navigation in the tree should work

vinokurig commented:
@svenefftinge 
>- Using the dockpanel doesn't seem right, as we have an odd invisible tabbar on the top.

If I add some css styles to the panel to remove those tabbars, will it be ok for you?
>- the dirty-ness doesn't work. The preferences editor is just an editor so it should behave like one.
>- the tree should gain focus

Could you please explain that more detailed
> - keyboard navigation in the tree should work

agree

svenefftinge commented:
>> Using the dockpanel doesn't seem right, as we have an odd invisible tabbar on the top.

> If I add some css styles to the panel to remove those tabbars, will it be ok for you?

I am not sure what other bad sideeffects it could have. Did you try using the SplitPanel?

> the dirty-ness doesn't work. 

When the editor has changes the closing-cross in the tabbar should indicate its dirtyness by turning into a circle. (Try with a regular text editor) 
You should implement `Saveable` and delegate everything to the editor. https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/saveable.ts#L23

> tree should gain focus

On `activate` you should set the focus to first node of the tree. For one it is important to set focus somewhere in your widget when it gets activated, but also as a user I want to use the keyboard to navigate, without using mouse/trackpad.

vinokurig commented:
@svenefftinge 
I've fixed all your remarks.

svenefftinge commented:
package.json in preferences needs `    "@theia/editor": "^0.3.12",` dependency.

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,237 @@
+        this.dirty = false;
+        super.onCloseRequest(msg);
+    }
+
```
We need to set the focus (`treewidget.activate`) during `onActivateRequest`.
```
    protected async onActivateRequest(msg: Message): Promise<void> {
        this._treeWidget.activate();
        super.onActivateRequest(msg);
    }
```

svenefftinge replied:
To clarify: Currently the focus is only set on first attach. If you add the method above it works properly on subsequent activation, too.

vinokurig replied:
done

svenefftinge commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,237 @@
+        this.initializeModel();
+        super.onAfterAttach(msg);
+    }
+
```
I think in this case it would make sense to allow using a simple click to trigger the same action.

svenefftinge replied:
I mean I would like to open the 'value options' with the left click.

vinokurig replied:
done

svenefftinge commented:
The build is still failing due to https://github.com/theia-ide/theia/pull/1983#issuecomment-403606183

vinokurig commented:
@svenefftinge fixed the build

svenefftinge commented:
It still fails

akosyakov commented on packages/core/src/browser/saveable.ts:
```diff
@@ -85,6 +85,7 @@ export namespace Saveable {
                         throw e;
                     }
                 } finally {
+                    widget.close = close;
```
could you elaborate on this change?

vinokurig replied:
If the widget's constructor is not called when the widget is reopened, this will prevent adding new overrided  close handler each time when the widget is reopened. Removed that because don't need this fix after overriding dispose() in the preferences widgets.

akosyakov commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,258 @@
+import { Emitter, Event, MaybePromise } from '@theia/core';
+import { Widget } from "@phosphor/widgets";
+
+export abstract class PreferencesContainer extends SplitPanel implements Saveable {
```
If the widget relies on another `Saveable` then it should implement `SaveableSource` instead.

vinokurig replied:
In this case `savable` is retrieved immediately after widget creation, but editor widget is initialized asynchronously.

vinokurig commented:
@svenefftinge The build is successful now, but the tests failed because of the 500 error: 
```
node-extension-server search:
     Error: 500: Internal Server Error for https://api.npms.io/v2/search?q=keywords%3Atheia-extension%20filesystem%20scope%3Atheia
      at Request._callback (src\node\npms.ts:15:83)
      at Request.self.callback (C:\projects\theia\node_modules\request\request.js:185:22)
      at Request.<anonymous> (C:\projects\theia\node_modules\request\request.js:1157:10)
      at IncomingMessage.<anonymous> (C:\projects\theia\node_modules\request\request.js:1079:12)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)
```

vinokurig commented:
@svenefftinge @akosyakov Any updates on this?

svenefftinge commented:
The menu creation should directly use phosphor, so doesn't need to side effects on our global command registry. 
```ts
import { injectable } from "inversify";
import { BrowserMainMenuFactory } from "@theia/core/lib/browser/menu/browser-menu-plugin";
import { Menu } from "@phosphor/widgets";
import { CommandRegistry } from "@phosphor/commands";
import { PreferenceProperty } from "@theia/core/lib/browser";

@injectable()
export class PreferencesMenuFactory extends BrowserMainMenuFactory {

    createPreferenceContextMenu(id: string, property: PreferenceProperty, execute: (property: string, value: any) => void): Menu {
        const commands = new CommandRegistry();
        const menu = new Menu({ commands });
        if (property) {
            const enumConst = property.enum;
            if (enumConst) {
                enumConst.forEach(enumValue => {
                    const commandId = id + '-' + enumValue;
                    if (!commands.hasCommand(commandId)) {
                        commands.addCommand(commandId, {
                            label: enumValue,
                            execute: () => execute(id, enumValue)
                        });
                        menu.addItem({
                            type: 'command',
                            command: commandId
                        });
                    }
                });
            } else if (property.type && property.type === 'boolean') {
                const commandTrue = id + '-true';
                commands.addCommand(commandTrue, {
                    label: 'true',
                    execute: () => execute(id, 'true')
                });
                menu.addItem({
                    type: 'command',
                    command: commandTrue
                });

                const commandFalse = id + '-false';
                commands.addCommand(commandFalse, {
                    label: 'false',
                    execute: () => execute(id, 'false')
                });
                menu.addItem({
                    type: 'command',
                    command: commandFalse
                });
            } else {
                const commandId = id + '-stringValue';
                commands.addCommand(commandId, {
                    label: 'Add Value',
                    execute: () => execute(id, property.default ? property.default : '')
                });
                menu.addItem({
                    type: 'command',
                    command: commandId
                });
            }
        }
        return menu;
    }
}
```

svenefftinge commented on packages/core/src/browser/menu/browser-menu-plugin.ts:
```diff
@@ -120,7 +120,7 @@ class DynamicMenuBarWidget extends MenuBarWidget {
  * A menu widget that would recompute its items on update
  */
-class DynamicMenuWidget extends MenuWidget {
+export class DynamicMenuWidget extends MenuWidget {
```
not needed when using phosphor menu directly.

vinokurig replied:
fixed

vinokurig commented:
@svenefftinge changed the  `PreferencesMenuFactory`

AlexTugarev commented:
Nice @vinokurig! 
I start reviewing this, too.

AlexTugarev commented:
I've noticed...
 - there is no feedback when I use the tree editor while the code editor part is dirty
  - we should warn in that case, no?
 - undo is not working in code editor
 - click should also select leaf nodes (not only show context menu)
  - it looks bogus that current item is not selected:
    <img width="645" alt="screen shot 2018-07-20 at 15 12 28" src="https://user-images.githubusercontent.com/914497/43004233-63381b08-8c2f-11e8-8ee1-20bea9481eb0.png">
 - (already mentioned) keyboard navigation is incomplete, I cannot activate the menu
 - it takes too long to the tooltips to appear (html `title`, we cannot fix it)
   - maybe okay for the beginning, but we should improve by showing another hover widget

AlexTugarev commented:
<img width="965" alt="screen shot 2018-07-20 at 15 23 49" src="https://user-images.githubusercontent.com/914497/43004744-fc74b3b6-8c30-11e8-9438-1ac56fe0a837.png">

Besides the functional aspects, I'd suggest to improve the widget to be more self-explaining. 

On the left hand side you see the available preferences with (kind of hidden) descriptions, but that's not necessarily clear for users.

On the right hand side you can place the preferences to overwrite the default User/Workspace prefs. Actually, it now gets important to know about the default.

AlexTugarev commented:
Also, why do we need to different preference widgets? I think it would be great to have one with an tabfolder for User and Workspace on the right hand side. Sometimes I need to search in both, in order to find why something is acting wrong.

Speaking of search, is it planned to add tree filter? I guess it's out of scope now.

vinokurig commented on packages/preferences/src/browser/preference-frontend-contribution.ts:
```diff
@@ -33,34 +43,53 @@ export namespace PreferenceCommands {
+
 @injectable()
-export class PreferenceFrontendContribution implements CommandContribution, MenuContribution {
+export class PreferencesFrontendContribution<T extends Widget> extends AbstractViewContribution<DockPanel> {
```
done

vinokurig commented on packages/preferences/src/browser/preference-frontend-module.ts:
```diff
@@ -18,19 +18,52 @@ import { ContainerModule, interfaces, } from 'inversify';
+    bind(WorkspacePreferencesWidgetFactory).toSelf().inSingletonScope();
+    bind(WidgetFactory).toDynamicValue(ctx => ctx.container.get(WorkspacePreferencesWidgetFactory)).inSingletonScope();
+
+    bind(UserPreferencesTreeWidget).toDynamicValue(ctx => createUserPreferencesTreeWidget(ctx.container));
```
fixed

vinokurig commented on packages/preferences/src/browser/preference-frontend-module.ts:
```diff
@@ -18,19 +18,52 @@ import { ContainerModule, interfaces, } from 'inversify';
+        createWidget: () => context.container.get<UserPreferencesTreeWidget>(UserPreferencesTreeWidget)
+    })).inSingletonScope();
+
+    bind(WorkspacePreferencesTreeWidget).toSelf();
```
fixed

vinokurig commented on packages/preferences/src/browser/preference-frontend-module.ts:
```diff
@@ -18,19 +18,52 @@ import { ContainerModule, interfaces, } from 'inversify';
-    bind(CommandContribution).toService(PreferenceFrontendContribution);
-    bind(MenuContribution).toService(PreferenceFrontendContribution);
+    bind(UserPreferencesFrontendContribution).toSelf().inSingletonScope();
+    bind(CommandContribution).toService(UserPreferencesFrontendContribution);
```
done

vinokurig commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,266 @@
+    private _editorWidget: EditorWidget;
+
+    protected constructor(protected readonly widgetManager: WidgetManager,
+                          @unmanaged() readonly id: string,
```
fixed

vinokurig commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,266 @@
+
+export class UserPreferencesContainer extends PreferencesContainer {
+
+    constructor(@inject(WidgetManager) protected readonly widgetManager: WidgetManager,
```
fixed

vinokurig commented on packages/preferences/src/browser/preference-menu-factory.ts:
```diff
@@ -0,0 +1,78 @@
+import { PreferenceProperty } from "@theia/core/lib/browser";
+
+@injectable()
+export class PreferencesMenuFactory {
```
Could you please point me what's wrong with this name.

AlexTugarev replied:
I think it should be either `preferences-menu-factory.ts` or `PreferenceMenuFactory `.

vinokurig replied:
done

vinokurig commented on packages/filesystem/package.json:
```diff
@@ -4,6 +4,7 @@
   "description": "Theia - FileSystem Extension",
   "dependencies": {
     "@theia/core": "^0.3.12",
+    "@theia/editor": "^0.3.12",
```
fixed

vinokurig commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,266 @@
+    protected readonly toDispose = new DisposableCollection();
+    private readonly onDirtyChangedEmitter: Emitter<void>;
+
+    private readonly _treeWidgetId: string;
```
fixed

AlexTugarev commented on packages/preferences/src/browser/preferences-widget.ts:
```diff
@@ -0,0 +1,266 @@
+    }
+}
+
+export abstract class PreferencesTreeWidget extends TreeWidget {
```
but that makes it not easy to find, as you are not naming the file after the content, cf. 
https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names

vinokurig replied:
done

vinokurig commented:
@AlexTugarev 

>- there is no feedback when I use the tree editor while the code editor part is dirty
>- we should warn in that case, no?

There is a dirty-state indicator in the widget's tab
>- undo is not working in code editor
>- click should also select leaf nodes (not only show context menu)
>- it looks bogus that current item is not selected:

Fixed that
>- (already mentioned) keyboard navigation is incomplete, I cannot activate the menu

Added menu activation by Enter key
>Besides the functional aspects, I'd suggest to improve the widget to be more self-explaining.
On the left hand side you see the available preferences with (kind of hidden) descriptions, but that's not necessarily clear for users.

Since the widget is based on the tree I have no idea how to view the description directly.

> On the right hand side you can place the preferences to overwrite the default User/Workspace prefs. Actually, it now gets important to know about the default.

It is already works in that way. Some preferences may not contain the default value, so empty value would be added.

> Also, why do we need to different preference widgets? I think it would be great to have one with an tabfolder for User and Workspace on the right hand side. Sometimes I need to search in both, in order to find why something is acting wrong.

I prefer to have separate actions to avoid having double tab area above the widget @svenefftinge WDYT?

> Speaking of search, is it planned to add tree filter? I guess it's out of scope now.
I'll try to implement it in this PR

AlexTugarev commented:
>> Besides the functional aspects, I'd suggest to improve the widget to be more self-explaining.

>> On the left hand side you see the available preferences with (kind of hidden) descriptions, but that's not necessarily clear for users.

> Since the widget is based on the tree I have no idea how to view the description directly.

Can you have a look at the rendering of the Call Hierarchy tree? It renders with different colors. Maybe we can render the default values grayish after a dash? 

>> On the right hand side you can place the preferences to overwrite the default User/Workspace prefs. Actually, it now gets important to know about the default.

> It already works in that way. Some preferences may not contain the default value, so empty value would be added.

What I meant is, it's too late then to decide if I want to override a value. If I'd know the default value before, I would not make a change. Can we highlight the default value in the context menu? If it's already like that, I didn't notice.

AlexTugarev commented:
>> Also, why do we need to different preference widgets? I think it would be great to have one with an tabfolder for User and Workspace on the right hand side. Sometimes I need to search in both, in order to find why something is acting wrong.

> I prefer to have separate actions to avoid having double tab area above the widget @svenefftinge WDYT?

Having a tab folder above the code editor part sounds handy in comparison to opening two instances and searching twice for the same keys. Have a look at vscode and vscode-insiders.

AlexTugarev commented:
>> there is no feedback when I use the tree editor while the code editor part is dirty
>> we should warn in that case, no?

> There is a dirty-state indicator in the widget's tab

That's not enough. When I tried, I did changes in code editor and was able to select items from tree, but the code editor was not in sync. It's not obvious which is the truth then, as I could (once fixed) undo, but the resource is modified, right? I think it's less confusing to not allow tree-based modifications while the editor is dirty.

vinokurig commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,276 @@
+import { Widget } from "@phosphor/widgets";
+import { EditorBasedSplitPanel } from "@theia/editor/lib/browser/editor-widget";
+
+export abstract class PreferencesContainer extends EditorBasedSplitPanel implements Saveable {
```
@AlexTugarev 
While reworking the user and workspace preferences widgets to single one I've faced with a problem: What to do with the dirty-state indicator? Since the widgets are placed in the common panel they don't have the close icon.
![screenshot-localhost-3000-2018 07 25-14-44-55](https://user-images.githubusercontent.com/7668752/43199042-5990b130-9019-11e8-9643-fea10d2536fc.png)

AlexTugarev replied:
@vinokurig I meant to have a tab folder for the code editor part only.

AlexTugarev replied:
try to save on change between both parts. e.g. having changes in user preferences and switching to workspace preferences would save user preferences.

AlexTugarev replied:
the tree part should stay stable, I think. therefore, please try with the tab folder above the code editors.

vinokurig replied:
So dirty indicator is missed, right?

vinokurig commented:
@AlexTugarev gathered User and workspace preferences into one widget:
![screenshot-localhost-3000-2018 07 27-15-17-48](https://user-images.githubusercontent.com/7668752/43320237-73f38fc8-91b0-11e8-8bba-922afafc1731.png)

vinokurig commented:
Added Warning notification if the editr(s) is/are in the dirty state
![screenshot-localhost-3000-2018 07 27-15-31-10](https://user-images.githubusercontent.com/7668752/43320764-3f0b245e-91b2-11e8-885a-680d0d026cf0.png)

vinokurig commented:
Added colorizing preference node if it is stored in the preferences files:
![screenshot-localhost-3000-2018 07 27-17-06-24](https://user-images.githubusercontent.com/7668752/43326414-7e608c3e-91c1-11e8-89e2-ea8ab6e4f6d8.png)

vinokurig commented:
@slemeur Since the last demo of this widget some new features have been added:

- Keyboard navigation
- Description is present directly near the preference node
- Default (if present) / stored preference is checked in the context menu
- Stored preferences are colored in yellow
- User/Workspace preferences are combined in single widget

![ezgif-1-842358b1fa](https://user-images.githubusercontent.com/7668752/43381641-efcece4e-93dd-11e8-88b9-2e7fa21c666b.gif)

Do you have something to add?

vinokurig commented:
@AlexTugarev 
Looks like I've fixed all your remarks. As for tree filter I think it could be done in the next iteration. I think the find mechanism from the File Navigator widget should be moved to some common place to make it usable by other widgets e.g. Preferences widget.

AlexTugarev commented:
@vinokurig, that's a lot of progress! Looking forward to try it, and hope to find the time today.


> Editable Feedback
> When the user see the list of preferences, there is no visual feedback shown to the user so he can understand how to use the widget. In your first version, you were displaying a small "Edit" icon on hover. I think that was useful and much better guiding the user into what he should be doing.

This special editor is not very self-explaining yet. I'd suggest to not hide the explanation in a hover, but add a decent header section which explains the editor.

svenefftinge commented:
@vinokurig great progress!
I think we can merge this as a start, since it is much better than the current state. Additional improvements and features could be handled subsequently.
Maybe you could remove the two slashes in the descriptions.

svenefftinge commented:
I second most of @slemeur's comments. Especially
 - seeing default values
 - indicating edit action on hover

I'd also like to filter/search. But I think that can all be done subsequentially. @AlexTugarev WDYT?

AlexTugarev commented:
Esp. the filtering it some improvement, that can be done later. 

If possible, it would be nice to improve description and default value, if that's not much work.

It seems the code changed a lot since I've scanned it last week. Would be nice to have a quick look. 

Otherwise, this is very nice!

vinokurig commented:
@AlexTugarev @svenefftinge @slemeur 

- Removed the slashes before the description
- Added number of preferences in the group section
- Added default / stored value directly after the preference name
- Make the preference name **bold** if it is stored in the preference files

![screenshot-localhost-3000-2018 07 30-15-51-59](https://user-images.githubusercontent.com/7668752/43398710-8b4d383c-9411-11e8-8ebd-4ebf064d9d62.png)

vinokurig commented:
@slemeur 
>Propertie's Description
I feel the way the description of the properties are display overwhelming.
As there is no soft-wrap, the user has to scroll horizontally to see the full description, which is not delivering a nice UX.
What other options do you have with the tree widget? Would you be able to display the description on a separate line with soft-wrap?

Adding a separate line to the tree node requires changes in `TreeDecoration` class. I think it could be done in the next iteration
>Editable Feedback
When the user see the list of preferences, there is no visual feedback shown to the user so he can understand how to use the widget. In your first version, you were displaying a small "Edit" icon on hover. I think that was useful and much better guiding the user into what he should be doing.

According to https://github.com/theia-ide/theia/pull/1983#issuecomment-408788818 My proposal is to add some description related to the widget's usage to the tree root node e.g:
![screenshot-localhost-3000-2018 07 30-16-24-52](https://user-images.githubusercontent.com/7668752/43400069-60ed89bc-9415-11e8-80b4-960493dafbb8.png)

>If user is adding a property on the right editor, will it colorized that property in yellow in the tree?

The property would be bold if it is added to a preference editor manually

vinokurig commented:
@slemeur 
>On your screenshot, you still have properties which have no default, like "lineNumbers". As I can see the line numbers in the editor's gutter, I think it has a default value. Why is it not displayed?

Some preferences may not have default values and if such values are not configured in the preferences files, no value will be shown in the tree.
>Can we change the first item from the tree to "Configure your preferences by clicking on the property." instead of "Apply the preference to selected preferences files"

Since we have several ways to apply the value I think `clicking on the property` would be not the full truth.

vinokurig commented:
@slemeur 
>> Some preferences may not have default values and if such values are not configured in the preferences files, no value will be shown in the tree.

>Why would they be no default value for `LineNumbers` ? They are displayed by default in the editor, so I'm expecting there is a default value.

All available preferences are red from `PreferencesSchema`, the `editor.lineNumbers` preference doesn't have a `default` field:
https://github.com/theia-ide/theia/blob/1e3f8dbb0915cc8ee2f47ddffa1f7e597cb3f992/packages/editor/src/browser/editor-preferences.ts#L36-L42

Looks like the Monaco editor controls the line-numbers by itself, if the preference is not stored.

>>Since we have several ways to apply the value I think clicking on the property would be not the full truth.

>Does the interaction always start with a click on the item?

The menu can be opened either by mouse click, either by ENTER key

slemeur commented:
@vinokurig : ok, got it.

So, I understand the technical implications. Still, for a user is looks improper and not accurate to say that there is no "default value" for the editor's `LineNumber` property, while there is definitely one used by the system.

When you display the preferences - user is expecting to have a single consolidated view of all the properties. He does not really care if it is Monaco that his defaulting it or another piece of the product.

What solution can you propose for that?

slemeur commented:
@vinokurig : Ok.

As far as it gets fixed, I'm fine with any resolution path you prefer to take 👍 ! 

Could you create the different issues that are related to the comments we had in this PR and make sure they are taken into account for the upcoming sprint? (cc @vparfonov)

svenefftinge commented:
Thanks! Could you rebase and squash?

vinokurig commented:
@svenefftinge done!

AlexTugarev commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -69,7 +69,13 @@ export class EditorManager extends WidgetOpenHandler<EditorWidget> {
     protected updateActiveEditor(): void {
         const widget = this.shell.activeWidget;
-        this.setActiveEditor(widget instanceof EditorWidget ? widget : undefined);
+        if (widget instanceof EditorWidget) {
```
@vinokurig why is it needed to extend editor manager that way? haven't seen that change.

would it have other implications? 
/cc @akosyakov

vinokurig replied:
Preferences editors are nested in the SplitPanel. If Focus changed `widget` will be SplitPanel so need to get editor from the SplitPanel

AlexTugarev replied:
Is it only because of the focus? Have you tried to set focus (or to activate) on the code editors when the host widget (SplitPanel) is activated?

vinokurig replied:
Yes I have, When performing any action from one of the preference editor `current` editor is retrieved but `current = undefined`  in that case

AlexTugarev replied:
Ah! Is that here because of keybindings?

AlexTugarev replied:
I'll try to look in this. Maybe we find a more generic way. I think it might be interesting for other parts to embed a code editor, without extending core, right?

svenefftinge replied:
See my other comment. I think introducing a provider is fine in core.

vinokurig replied:
done

AlexTugarev commented on packages/preferences/src/browser/preference-tree-container.ts:
```diff
@@ -0,0 +1,42 @@
+import { createTreeContainer, TreeDecoratorService, TreeWidget} from "@theia/core/lib/browser";
+import { PreferencesTreeWidget } from "./preferences-tree-widget";
+
+function createUserPreferencesTreeContainer(parent: interfaces.Container): Container {
```
leftover: `createUserPreferencesTreeContainer ` `->` `createPreferencesTreeContainer `

vinokurig replied:
fixed

AlexTugarev commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,307 @@
+let dirty = false;
+let currentEditor: EditorWidget;
+let preferencesScope: PreferenceScope;
+const onDirtyChangedEmitter = new Emitter<void>();
```
please check the life-cycle. it looks like, this will leak.

also, please move to a `namespace PreferencesContainer` if it's unrelated, or add to class if, it's a property.

vinokurig replied:
fixed

AlexTugarev commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,307 @@
+    protected treeWidget: TreeWidget;
+
+    get dirty(): boolean {
+        return dirty;
```
just asking: what exactly is `dirty` in that case? we're embedding two code editors, right?

vinokurig replied:
The whole widget is dirty, the widget is kind of an editor

AlexTugarev replied:
the widget is a saveable, right? as such it delegates to the embedded code editors, like this
```
get dirty(): boolean {
  return editor1.dirty || editor2.dirty;
}
```

vinokurig replied:
fixed

vinokurig commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,307 @@
+                this.properties[node.id],
+                (property, value) => {
+                    if (dirty) {
+                        this.messageService.warn('Preferences editor(s) has/have unsaved changes');
```
No, it won't. Should it save on editor switch?

AlexTugarev replied:
It would be easier to manage the dirty state. Also you (as user) would not need to think about it.

vinokurig replied:
OK I will improve it

vinokurig commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,307 @@
+    autoSave: "on" | "off";
+    save: () => MaybePromise<void>;
+    onDirtyChanged: Event<void>;
+    protected readonly toDispose = new DisposableCollection();
```
fixed

vinokurig commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+export class PreferencesTreeWidget extends TreeWidget {
```
This class is not injected to any class

vinokurig commented:
@AlexTugarev I've fixed all your remarks, could you please take a look?

svenefftinge commented on packages/editor/src/browser/editor-widget.ts:
```diff
@@ -18,6 +18,11 @@ import { SelectionService } from '@theia/core/lib/common';
 import { TextEditor } from './editor';
+import { SplitPanel } from '@phosphor/widgets';
+
+export abstract class EditorBasedSplitPanel extends SplitPanel {
```
This should be an interface independent of the SplitPanel:
```
export interface EditorWidgetProvider {
    getEditor(): EditorWidget;
}
export namespace EditorWidgetProvider {
    export function is(provider: object | undefined): provider is EditorWidgetProvider {
        return !!provider && 'getEditor' in provider;
    }
}
```

vinokurig replied:
@svenefftinge Done!

svenefftinge commented:
+1 @AlexTugarev are you fine with it, too?

AlexTugarev commented:
I’ll have a look shortly and test a bit. Looks good so far.

akosyakov commented on packages/preferences/src/browser/preferences-tree-widget.ts:
```diff
@@ -0,0 +1,352 @@
+            }
+            this.currentEditor = editor;
+            const currentWidget = this.shell.currentWidget;
+            this.shell.currentChanged.emit({
```
One can only listen to current and active widget changes, only the shell can emit. We should hide this API by replacing signals with events.

svenefftinge replied:
Agreed. @vinokurig Why do we need to set the inner editor as current, here?

vinokurig replied:
`ApplicationShell` has a focus tracker that fires editor-changed events, but it didn't work if editor changed inside a widget, so need to fire the event manually.

svenefftinge replied:
I created a PR on your fork https://github.com/vinokurig/theia/pull/1
@akosyakov Could you have alook, too?

AlexTugarev commented:
Yeah!🎉
